src/scheduler/worker.rs

```diff
@@ -52,11 +52,36 @@ impl<VM: VMBinding> GCWorkerShared<VM> {
 
 /// Used to synchronize mutually exclusive operations between workers and controller,
 /// and also waking up workers when more work packets are available.
-/// NOTE: All fields are public in order to support the complex control structure
-/// in `GCWorkScheduler::poll_slow`.
 pub(crate) struct WorkerMonitor {
-    pub sync: Mutex<WorkerMonitorSync>,
-    pub cond: Condvar,
+    /// The synchronized part.
+    sync: Mutex<WorkerMonitorSync>,
+    /// This is notified when new work is made available for the workers.
+    /// Particularly, it is notified when
+    /// -   `sync.worker_group_state` is transitioned to `Working` because
+    ///     -   some workers still have designated work, or
+    ///     -   some sentinel work packets are added to their drained buckets, or
+    ///     -   some work buckets are opened, or
+    /// -   any work packet is added to any open bucket.
+    /// Workers wait on this condvar.
+    work_available: Condvar,
+    /// This is notified when all workers parked.
+    /// The coordinator waits on this condvar.
+    all_workers_parked: Condvar,
+}
+
+/// The state of the worker group.  The worker group alternates between the `Sleeping` and the
+/// `Working` state.  Workers execute work packets in the `Working` state, but once workers entered
+/// the `Sleeping` state, they can not continue until the coordinator explicitly transitions the
+/// state back to `Working` after it found more work for workers to do.
```

Comment:
The cause and effect is the opposite.  The state dictates the behaviour instead of behaviour determining the state.  The `Working` state *allows* workers to execute and *forbids* the coordinator from opening new buckets, while the `Sleeping` state *forbids* workers from executing packets and *allows* the coordinator to open buckets.